### PR TITLE
Check that all alerts make it into loki

### PR DESCRIPTION
**Pull request by dstathis**

### CODE REVIEW

**Review by simskij — APPROVED** [review]

File: tests/integration/test_loki_push_api.py [hunk 1]
```diff
@@ -72,8 +72,8 @@ async def test_rule_files_ingested_by_loki(ops_test):
     await ops_test.model.wait_for_idle(status="active", timeout=1000)
 
     # now, make sure rules are present
+    expected = {"HighThroughputLogStreams", "HighPercentageError", "http-credentials-leaked"}
     response = await client.rules()
     assert (await client.rules()).items() > {}.items()
-    alert_group = next(iter(response.values()))[0]
-    assert "name" in alert_group
-    assert "rules" in alert_group
+    alerts = [group["rules"][0]["alert"] for group in next(iter(response.values()))]
```
Comment (sed-i) [inline tests/integration/test_loki_push_api.py:78]: Now that we're checking for presence of alert names by name, the code must be more sophisticated than `rules[0]` - a rules group is a list of alerts. I.e. in a single file there can be multiple group names under `groups` and multiple alert names under `rules`.

**Review by sed-i — CHANGES_REQUESTED** [review]

**Review by Abuelodelanada — APPROVED** [review]

File: tests/samples/loki_alert_rules/upstream_example.rule [hunk 2]
```diff
@@ -0,0 +1,23 @@
+groups:
```
[the remark this answers is not in the document]
Comment (dstathis) [reply tests/samples/loki_alert_rules/upstream_example.rule]: I rebased. Not sure why it's still showing up in the diff here but it should merge correctly.

File: tests/integration/test_loki_push_api.py [hunk 3]
```diff
@@ -72,8 +72,8 @@ async def test_rule_files_ingested_by_loki(ops_test):
...
+    assert set(alerts) == expected
```
[the remark this answers is not in the document]
Comment (dstathis) [reply tests/integration/test_loki_push_api.py:79]: We do have duplicates which is why I did set. Sorted would be better.

[the remark this answers is not in the document]
Comment (dstathis) [reply tests/integration/test_loki_push_api.py]: Changed to sorted.

File: tests/integration/test_loki_push_api.py (hunk 1 above)
Comment (dstathis) [reply tests/integration/test_loki_push_api.py]: Done.

File: tests/integration/test_loki_push_api.py [hunk 4]
```diff
@@ -72,8 +72,17 @@ async def test_rule_files_ingested_by_loki(ops_test):
     await ops_test.model.wait_for_idle(status="active", timeout=1000)
 
     # now, make sure rules are present
```
Comment (sed-i) [inline tests/integration/test_loki_push_api.py:74]:
```suggestion
    # now, make sure rules are present
    # in the samples dir, the same alert name "HighThroughputLogStreams" appears
    # under different files/groups three times and is therefore accounted here three times
```

File: tests/integration/test_loki_push_api.py [hunk 5]
```diff
@@ -72,8 +72,17 @@ async def test_rule_files_ingested_by_loki(ops_test):
...
+    expected = [
+        ["HighPercentageError"],
+        ["HighThroughputLogStreams"],
+        ["HighThroughputLogStreams"],
+        ["HighThroughputLogStreams"],
+        ["http-credentials-leaked"],
+    ]
     response = await client.rules()
     assert (await client.rules()).items() > {}.items()
```
Comment (sed-i) [inline tests/integration/test_loki_push_api.py:83]: Feel free to remove this one - it is now redundant.

File: tests/integration/test_loki_push_api.py [hunk 6]
```diff
@@ -72,8 +72,17 @@ async def test_rule_files_ingested_by_loki(ops_test):
...
-    alert_group = next(iter(response.values()))[0]
-    assert "name" in alert_group
-    assert "rules" in alert_group
+    alerts = [
+        sorted([rule["alert"] for rule in group["rules"]])
+        for group in next(iter(response.values()))
```
Comment (sed-i) [inline tests/integration/test_loki_push_api.py:86]: I don't remember the structure coming out in the response, but shouldn't it be
```suggestion
        for group in response.values()
```

**Review by sed-i — APPROVED** [review]

File: tests/integration/test_loki_push_api.py (hunk 6 above)
Comment (dstathis) [reply tests/integration/test_loki_push_api.py:86]: No values returns a list like object with one item that can't be subscripted. (Also this is how it was before)

File: tests/integration/test_loki_push_api.py (hunk 5 above)
Comment (dstathis) [reply tests/integration/test_loki_push_api.py:85]: Remove what? Do you mean `assert (await client.rules()).items() > {}.items()`

Comment (sed-i) [reply tests/integration/test_loki_push_api.py:85]: Yes, it is obviated by your addition.

**Pull request merged**